// flow-static-land/lib/Data.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var Data1 = exports.Data1 = function Data1(value0) {
  _classCallCheck(this, Data1);

  this.value0 = value0;
};

var Data2 = exports.Data2 = function Data2(value0, value1) {
  _classCallCheck(this, Data2);

  this.value0 = value0;
  this.value1 = value1;
};

var Data3 = exports.Data3 = function Data3(value0, value1, value2) {
  _classCallCheck(this, Data3);

  this.value0 = value0;
  this.value1 = value1;
  this.value2 = value2;
};

var Data4 = exports.Data4 = function Data4(value0, value1, value2, value3) {
  _classCallCheck(this, Data4);

  this.value0 = value0;
  this.value1 = value1;
  this.value2 = value2;
  this.value3 = value3;
};

var Data5 = exports.Data5 = function Data5(value0, value1, value2, value3, value4) {
  _classCallCheck(this, Data5);

  this.value0 = value0;
  this.value1 = value1;
  this.value2 = value2;
  this.value3 = value3;
  this.value4 = value4;
};
